File: stdlib/src/stdlib/platform/Lists.js
var Builtins = require("../../builtins");
var Options = require("../options");
var Sequences = require("../Sequences");

function listToSequence(elements) {
    return listIndexToSequence(elements, 0);
}

function listIndexToSequence(elements, index) {
    return Sequences.Sequence({
        next: function() {
            if (index < elements.length) {
                return Options.some(Sequences.SequenceItem({
                    head: elements[index],
                    tail: listIndexToSequence(elements, index + 1)
                }));
            } else {
                return Options.none;
            }
        }
    });
}

function sequenceToList(sequence) {
    var result = [];
    
    while (true) {
        var item = sequence.next();
        if (Builtins.isType(item, Options.None)) {
            return result;
        } else {
            result.push(item.value.head);
            sequence = item.value.tail;
        }
    }
}

exports.listToSequence = listToSequence;
exports.sequenceToList = sequenceToList;

File: backend-python/src/main/resources/org/shedlang/compiler/backends/python/modules/stdlib/platform/Lists.py
from .. import Options, Sequences


def list_to_sequence(elements):
    return _list_index_to_sequence(elements, 0)


def _list_index_to_sequence(elements, index):
    def next_item():
        if index < len(elements):
            return Options.some(
                Sequences.SequenceItem(
                    head=elements[index],
                    tail=_list_index_to_sequence(elements, index + 1),
                )
            )
        else:
            return Options.none
    
    return Sequences.Sequence(
        next=next_item,
    )


def sequence_to_list(sequence):
    result = []
    
    while True:
        item = sequence.next()
        if isinstance(item, Options.None_):
            return result
        else:
            result.append(item.value.head)
            sequence = item.value.tail
